[1.6] Prefer storage.k8s.io/v1beta1 for discovery

ncdc opened the pull request:
Prefer storage.k8s.io/v1beta1 for discovery, as older clients might have
a hard time if they don't know about v1.

Fixes #45134 

cc @liggitt @jsafrane @soltysh @enj @kubernetes/sig-storage-pr-reviews @kubernetes/sig-api-machinery-pr-reviews

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45138)
<!-- Reviewable:end -->

deads2k commented:
> Prefer storage.k8s.io/v1beta1 for discovery, as older clients might have
> a hard time if they don't know about v1.

Our clients don't fallback to a version they recognize?  That's a thing we'll want fixed upstream in 1.7 I'd think.

deads2k commented on pkg/apis/storage/install/install.go:
```diff
@@ -38,7 +38,21 @@ func Install(groupFactoryRegistry announced.APIGroupFactoryRegistry, registry *r
+				`cluster/update-storage-objects.sh` prior to upgrading to 1.7. This will update all
+				storageclasses so they're stored in v1 format in etcd.
+			*/
 			VersionPreferenceOrder:     []string{v1beta1.SchemeGroupVersion.Version, v1.SchemeGroupVersion.Version},
```
hmmm.  This doesn't seem like the right spot.  I thought we chose storage version distinctly from API interaction version. Even if we don't at this very moment, it seems like we should.

ncdc replied:
Yes, I agree it should be explicit and done elsewhere.

ncdc commented:
> Our clients don't fallback to a version they recognize? That's a thing we'll want fixed upstream in 1.7 I'd think.

For the moment I'm going off of what @liggitt said happened in 1.5. I am hoping to get some time to test a 1.5 kubectl against this PR.

Also, when working on `test-update-storage-objects.sh`, the only way I could get the last part of the test to pass was to delete the cached discovery information before running `kubectl get`. It has storage/v1beta1 and storage/v1 cached, but the only version available in the apiserver at this point is storage/v1. `kubectl` uses its cached discovery information and tries to get storage/v1beta1, while fails, and it then gives up. (https://github.com/kubernetes/kubernetes/pull/45138/files#diff-8dc3ac3856b6ae7bb59360438f5b4fe6R238)

ncdc commented:
Ok, I tested kubectl-1.5 against master, where v1 is both the preferred storage version and preferred discovery version. The only issue I ran into was trying to run `kubectl create -f foo.yaml` where foo.yaml contained a v1beta1 storageclass. It tried to post to v1, and got back an error that it couldn't process a v1beta1 storageclass because it needed to be v1. Everything else I tried (get, describe, create yaml with v1) appeared to work ok.

@deads2k suggested the following timeline:

|Kube Version|Available storage.k8s.io Versions|Encoding Version|Status|
|-|-|-|-|
|1.5|v1beta1|v1beta1|ok|
|1.6|v1, v1beta1|v1beta1|ok|
|1.7|v1, v1beta1|v1|ok|
|1.8|v1|v1|tbd|

We need to add support for storageclasses to `cluster/update-storage-objects.sh` and a test for it. This PR does those 2 things for release-1.6. I'll make a similar PR for master. Cluster admins must run this script prior to upgrading to Kubernetes 1.8, assuming we agree that we can drop storage.k8s.io/v1beta after 1.7.

enisoc commented:
Is there a reason this directly targets the `release-1.6` branch, rather than `master`?

sttts commented:
@enisoc in master (which becomes 1.7 in a few months) we can advertise v1 as the preferred version because kubectl one version back (1.6) knows about v1, a kubectl from 1.5 does not. To avoid confusing a kubectl from 1.5 with a 1.6 server we want this change.

enisoc commented:
@sttts Thanks for the explanation. I've approved it for cherrypick, but since it's not a copy of something from master, I will defer to the regular owners for LGTM+Approval.

ncdc committed:
Prefer storage.k8s.io/v1beta for discovery

Prefer storage.k8s.io/v1beta for discovery, as older clients might have
a hard time if they don't know about v1.

ncdc commented:
Ok, I've removed the changes to the scripts that update storage objects because admins must wait until 1.7 to do that. I will be opening a PR against master for that shortly.

ncdc commented:
#46116 has the update script for master

enisoc commented:
Are you interested in getting this in for 1.6.5 (scheduled Jun 14)? If so, please get reviewers to lgtm/approve.

enisoc commented:
Since there doesn't seem to be interest, I'm closing this to keep the cherrypick queue clean. Please reopen when you resume work on this.

ncdc commented:
@enisoc sorry, I was on vacation last week. As best I can tell, if we do not merge this, a 1.5 kubectl won't be able to `kubectl create` something containing a v1beta1 StorageClass when talking to a 1.6.x apiserver, but that was the extent of the compatibility issues that I could find.

enisoc commented:
@ncdc Thanks for the update. I'll reopen, but it still needs LGTM+Approval from code owners.

lavalamp commented:
Given that 1.7 is released now, is this actually worth adding to 1.6?

ncdc commented:
@lavalamp I dunno. This is mostly a bandaid to allow kubectl 1.5 to create v1beta1 storageclasses against a 1.6 server. I think the better fix, if we care, is to figure out why kubectl 1.5 ignores the Kind in the file passed to `create -f` and instead tries to POST to the group's preferred version (i.e. my file has storage.k8s.io/v1beta1, but kubectl POSTs to storage.k8s.io/v1, and that fails).

ncdc commented:
Given that no one is complaining about this, I'm closing it. We can reopen if needed.
